ref(discover) Simplify decodeScalar() and decodeList()

markstory opened the pull request:
Instead of returning string|undefined we can return '' and have a fallback parameter that lets us remove a bunch of `??` usage.

markstory committed:
ref(discover) Simplify decodeScalar() and decodeList()

Instead of returning string|undefined we can return '' and have
a fallback parameter that lets us remove a bunch of `??` usage.

k-fish reviewed (commented):
Nice 👏 . Looks like tsc error though.

markstory committed:
Merge branch 'master' into ref/simpler-decode

markstory committed:
Fix up a type error.

k-fish commented:
Look like it might be a real error?

markstory committed:
Merge branch 'master' into ref/simpler-decode

markstory committed:
Make changes more backwards compatible

Forcing the default value to '' changed how EventView worked in
significant ways. Using an overload gives the typing benefits and
removes the need to use `??` and `||` in general.

k-fish commented on src/sentry/static/sentry/app/utils/queryString.tsx:
```diff
@@ -51,26 +51,26 @@ export function appendTagCondition(
-  value: string[] | string | undefined | null
-): string | undefined {
+// This function has multiple signatures to help with typing in callers.
+export function decodeScalar(value: QueryValue): string | undefined;
```
Yeah these overloads make the types nicer.

markstory replied:
Using an overload only came to mind after struggling with the tests for a few minutes, and then seeing the fixes causing other failures. :blush:
